CLI: Modernize Storybook CLI with new init workflow, Clack UI, and Generator System

valentinpalkovic opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/25389
Closes https://github.com/storybookjs/storybook/issues/32043
Closes https://github.com/storybookjs/storybook/issues/32778
Closes https://github.com/storybookjs/storybook/issues/31880
Closes https://github.com/storybookjs/storybook/issues/31914

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

This PR represents a massive refactor and modernization of the Storybook CLI, primarily focused on the `init` command. The original initiate.ts file, a monolithic, 900+ line behemoth, was brittle, hard to maintain, and difficult to test.

This new architecture replaces it with a clean, command-based pipeline that is modular, testable, and far more extensible. We've also swapped our CLI's user interface from prompts to @clack/prompts, providing a significantly improved, modern, and interactive experience for users.

### Highlights of the Change
**Command-Based init Workflow: The entire init process is now orchestrated as a series of discrete, testable commands:**

- PreflightCheckCommand: Handles empty directory scaffolding and package manager setup.
- ProjectDetectionCommand: Detects the project type (e.g., React, Angular).
- FrameworkDetectionCommand: Determines the correct Storybook framework, builder, and renderer.
- UserPreferencesCommand: Handles all user interaction (New user? Install type?).
- GeneratorExecutionCommand: Runs the appropriate framework-specific generator.
- DependencyInstallationCommand: Collects all dependencies and installs them in a single step.
- AddonConfigurationCommand: Runs post-install configuration for addons like a11y and vitest.
- FinalizationCommand: Updates .gitignore and prints success messages.

**New Generator System:**
Introduced a generatorRegistry and a new GeneratorModule interface.

All framework generators (React, Angular, Next.js, Nuxt, etc.) have been refactored to this new, modular standard, separating metadata and configuration logic .

**Modern Clack UI:**
- Replaced `prompts`, `ora`, and `boxen` with `@clack/prompts` across the entire CLI (init, add, upgrade, automigrate) for a unified, beautiful, and interactive experience.


**Centralized Addon & Feature Logic:**
- AddonVitestService: A new, centralized service that manages all logic for the test feature. It handles compatibility validation, dependency collection, and configuration, ensuring logic is no longer duplicated between init and add .
- DependencyCollector: A new class that gathers all required dependencies before installation, allowing us to run a single npm install/yarn add at the end of the init process .

**a11y in `recommended`:** 
The test feature now also includes @storybook/addon-a11y by default, as it's a prerequisite for accessibility testing.

**New Next.js-Vite Support:**
- The Next.js generator now defaults to @storybook/nextjs-vite (Vite builder) for new projects. It intelligently detects custom webpack.config.js or .babelrc files and prompts the user to fall back to @storybook/nextjs (Webpack 5) if needed.

- Added a new nextjs-to-nextjs-vite automigration to help users move from the old Webpack-based setup .

**Rsbuild Support during initialization:**
- When initializing Storybook in an rsbuild-based project, we will set up your project properly if rsbuild is used with one of the supported renderers: lit, react, vue, vanilla js.


**Enhanced Logging & Error Handling:**
- All CLI commands now use the standardized logger and prompt from node-logger.
- Builder output (Vite) is now prefixed for clarity (e.g., [Vite] ...). Webpack was not adjusted yet due to its complex nature of logging items (status)
- All CLI commands (init, add, automigrate) will now reliably generate a `debug-storybook.log` file on failure, making debugging much easier.

**Introduced new StorybookError subclasses for CLI failures (e.g., AddonVitestPostinstallError)**

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

This is a massive change. Please test the init and add commands thoroughly.

`storybook init`
1. The "Golden Path":
  - `cd` into a new React + Vite project.
  - Run `npx storybook@<canary> init`.
  - ✅ Assert: You see the new Clack UI.
  - ✅ Assert: You are prompted "New to Storybook?" > "What configuration?". Select "Recommended".
  - ✅ Assert: It installs all recommended dependencies, including test (@storybook/addon-vitest, vitest, etc.) and a11y (@storybook/addon-a11y).
  - ✅ Assert: It prompts you to install Playwright binaries.
  - ✅ Assert: The final success message lists docs, test, a11y, onboarding.

2. Non-Interactive (--yes):
  - `cd` into a new React + Vite project.
  - Run `npx storybook@<canary> init --yes`.
  - ✅ Assert: No prompts appear.
  - ✅ Assert: The result is identical to the "Recommended" install (all features included).

3. Feature Selection:
  - `cd` into a new React + Vite project.
  - Run `npx storybook@latest init --features docs a11y`.
  - ✅ Assert: No prompts for "New User" or "Install Type" appear.
  - ✅ Assert: Only @storybook/addon-docs and @storybook/addon-a11y are installed. @storybook/addon-vitest and @storybook/addon-onboarding are not installed.

4. Next.js (Vite vs. Webpack):
  - `cd` into a clean Next.js project. Run init --yes.
  - ✅ Assert: It installs @storybook/nextjs-vite and vite.
  - Now, `rm -rf .storybook` and add a `webpack(config) { return config; }` block to next.config.js.
  - Run init interactively.
  - ✅ Assert: It must prompt you to choose between @storybook/nextjs-vite (default) and @storybook/nextjs (Webpack).

`storybook add`
- `cd` into a project without the test feature (e.g., init --features docs).
- Run `npx storybook@<canary> add @storybook/addon-vitest`.
- ✅ Assert: The postinstall script runs.
- ✅ Assert: You see the Clack UI prompt to install Playwright.
- ✅ Assert: It creates vitest.config.ts.
- ✅ Assert: If you also have addon-a11y installed, it should run the a11y-addon-test automigration to configure it.

`storybook automigrate`
- `cd` into a project using the old @storybook/nextjs (Webpack).
- Run npx storybook@latest automigrate nextjs-to-nextjs-vite.
- ✅ Assert: package.json and main.ts are updated from @storybook/nextjs to @storybook/nextjs-vite.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-32717-sha-7a16d827`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-32717-sha-7a16d827 sandbox` or in an existing project with `npx storybook@0.0.0-pr-32717-sha-7a16d827 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-32717-sha-7a16d827`](https://npmjs.com/package/storybook/v/0.0.0-pr-32717-sha-7a16d827) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`valentin/cli-init-rework`](https://github.com/storybookjs/storybook/tree/valentin/cli-init-rework) |
| **Commit** | [`7a16d827`](https://github.com/storybookjs/storybook/commit/7a16d827f5ca2f0a98ec372ed9bad4018d996a12) |
| **Datetime** | Wed Nov 19 10:25:34 UTC 2025 (`1763547934`) |
| **Workflow run** | [19498001525](https://github.com/storybookjs/storybook/actions/runs/19498001525) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=32717`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Modular init flow with pluggable generators and a dedicated addon-configuration command.

* **Improvements**
  * CLI: new --logfile and --loglevel options; "a11y" added to feature choices.
  * Smoother, more consistent logging and prompt UX across init, sandbox, and dev workflows.
  * Better automated handling for test integrations and dependency setup during initialization.

* **Tests**
  * Broader test coverage for init, generators, package managers, and migrations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

valentinpalkovic committed:
Update CircleCI configuration to use medium executor class for test-init-features job

valentinpalkovic committed:
Update CircleCI configuration to use medium+ executor class for test-init-features job

valentinpalkovic committed:
Revert (for debugging purposes

valentinpalkovic committed:
Refactor FrameworkDetectionCommand to enforce framework presence and clean up type definitions

- Changed framework property in FrameworkDetectionResult to be required instead of optional.
- Updated getFramework method to throw an error if a framework cannot be found, improving error handling.
- Removed unused type definition for Builder in project_types.ts to streamline code.

valentinpalkovic committed:
Refactor GeneratorExecutionCommand and baseGenerator to enhance framework handling and type definitions

- Updated generatorOptions to include framework and renderer properties for better configuration.
- Simplified the getFrameworkDetails function to improve clarity and error handling.
- Removed unused builder details logic to streamline the baseGenerator function.
- Adjusted types in GeneratorOptions to reflect the new structure.

valentinpalkovic committed:
Refactor baseGenerator to streamline framework package handling

- Removed the externalFrameworks function to simplify framework detection logic.
- Updated getFrameworkDetails to enhance clarity by directly applying the require wrapper on the framework package.
- Improved overall readability and maintainability of the baseGenerator code.

valentinpalkovic committed:
Enhance Angular generator to prompt for Compodoc usage

- Added a prompt to ask users if they want to use Compodoc for documentation in Angular projects.
- Improved user experience by providing information about the benefits of Compodoc.
- Updated the logic to determine Compodoc usage based on user input instead of a fixed feature flag.

valentinpalkovic committed:
Refactor generator execution and user preferences handling to improve addon configuration

- Updated the `doInitiate` function to pass `projectType` to user preferences and generator execution.
- Simplified the `executeGeneratorExecution` function to accept a single options object.
- Enhanced the `UserPreferencesCommand` to determine selected features based on project type.
- Adjusted addon configuration logic to utilize `extraAddons` from generator execution results.
- Improved overall code readability and maintainability by streamlining function signatures and logic.

valentinpalkovic committed:
Refactor GeneratorExecutionCommand to enhance type definitions and return structure

- Updated the `GeneratorExecutionResult` type to improve clarity and structure by formatting the return type.
- Adjusted the return statement in `executeProjectGenerator` to conditionally include `configDir` based on the generator result.
- Improved overall readability and maintainability of the `GeneratorExecutionCommand` class.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Install @storybook/addon-docs if docs should be installed

valentinpalkovic committed:
Refactor type definitions and logging in server-statics

- Updated the logging in `useStatics` to display relative paths for static directories instead of absolute paths, enhancing clarity.
- Removed unused `Builder` type imports and replaced them with `SupportedBuilder` in various files to improve type consistency and maintainability.

valentinpalkovic committed:
Update AddonManager tests to use SupportedBuilder constants

- Replaced string literals with SupportedBuilder constants in the AddonManager test cases for improved type safety and consistency.
- Enhanced clarity in the tests by using the defined types for builder options.

valentinpalkovic committed:
Refactor logging in server-statics to use project root for relative paths

- Updated the `useStatics` function to utilize `getProjectRoot` for calculating relative static directory paths, enhancing accuracy in logging.
- Changed the logger to use `CLI_COLORS` for improved visual consistency in log messages.

valentinpalkovic committed:
Refactor Compodoc logging in Angular generator for improved readability

- Consolidated the logging message in the `promptForCompoDocs` function to a single line for better clarity and maintainability.
- Enhanced user experience by providing a concise message about the benefits of using Compodoc for documentation in Angular projects.

valentinpalkovic committed:
Enable debug mode for init

valentinpalkovic committed:
Add yes option to GeneratorExecutionCommand for user confirmation

- Included the `yes` option in the `GeneratorExecutionCommand` to allow automatic confirmation for prompts.
- Enhanced the command's flexibility by enabling users to bypass interactive prompts based on their preferences.

valentinpalkovic committed:
WIP

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Refactor postinstall and logger integration across CLI components

- Updated postinstall function to accept logger and prompt from options.
- Enhanced ClackPromptProvider to manage task logging more effectively.
- Streamlined logger usage in add and postinstallAddon functions.
- Removed redundant task messages in baseGenerator for cleaner output.

valentinpalkovic committed:
Refactor logger messages for consistency and clarity

- Removed leading arrows from logger messages in various files to standardize output format.
- Introduced ConsoleLogger and StyledConsoleLogger classes for enhanced logging functionality.
- Added new console logger methods to improve logging capabilities and maintainability.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Enhance GeneratorOptions type and streamline baseGenerator parameters

- Added GeneratorOptions type to the GeneratorExecutionCommand for improved type safety.
- Simplified parameter destructuring in baseGenerator by removing unused properties.
- Updated frameworkPreviewParts handling in baseGenerator to utilize options directly.

valentinpalkovic committed:
Refactor framework configuration in baseGenerator for simplicity

- Streamlined the framework configuration in baseGenerator by removing unnecessary object wrapping around the framework name.
- Improved readability and maintainability of the code by directly assigning the frameworkPackagePath.

valentinpalkovic committed:
Remove ConfigGenerationService and its associated tests

- Deleted the ConfigGenerationService implementation and its test file to streamline the codebase.
- This change eliminates unused code and focuses on maintaining only necessary components for configuration generation.

valentinpalkovic committed:
Add logger and prompt expectations in AddonConfigurationCommand tests

- Updated tests for AddonConfigurationCommand to include expectations for logger and prompt objects.
- Enhanced test coverage by ensuring that these properties are correctly passed during the command execution.

valentinpalkovic committed:
Refactor feature handling in CLI components to use Feature enum

- Updated various CLI command files to replace string-based feature handling with a new Feature enum for improved type safety and consistency.
- Modified the copyTemplateFiles function to utilize Set<Feature> instead of string arrays for feature management.
- Removed deprecated GeneratorFeature type and associated tests to streamline the codebase.
- Enhanced overall maintainability and readability of the code by standardizing feature references.

valentinpalkovic committed:
Add Feature enum to exports and clean up ConfigGenerationService exports

- Added 'Feature' to the exported types in exports.ts for enhanced type safety.
- Removed unused exports related to ConfigGenerationService in index.ts to streamline the codebase and improve maintainability.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Refactor dependency management and introduce AddonService

- Replaced the DependencyCalculator with a new AddonService to manage Storybook addons more effectively.
- Updated GeneratorExecutionCommand and baseGenerator to utilize the AddonService for addon configuration based on selected features.
- Removed the DependencyCalculator and its associated tests to streamline the codebase.
- Added tests for the new AddonService to ensure correct functionality and feature handling.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Add logging for build and start processes in Angular Storybook builders

- Introduced logger to the build-storybook and start-storybook builders to provide feedback during the build and start processes.
- Added introductory log messages: "Building storybook" and "Starting storybook" to enhance user experience and debugging capabilities.

valentinpalkovic committed:
Enhance AddonService tests to include A11Y feature and verify addon configurations

- Updated tests to check for the addition of '@storybook/addon-vitest' and '@storybook/addon-a11y' when the respective features are enabled.
- Modified the AddonService implementation to ensure proper handling of the A11Y feature in addon configurations.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Merge branch 'next' into valentin/cli-init-rework

valentinpalkovic committed:
Refactor telemetry notification message formatting

- Replaced string concatenation with `ts-dedent` for improved readability.
- Combined multiple log statements into a single dedented message that includes information about anonymous telemetry collection and a URL for opting out.

valentinpalkovic committed:
Update CircleCI configuration to use `--loglevel=debug` for create-storybook command

- Changed the `--debug` flag to `--loglevel=debug` in both `.circleci/config.yml` and `.circleci/src/jobs/test-init-features.yml` to enhance logging during the Storybook initialization process.

valentinpalkovic committed:
Update CircleCI configuration to skip installation during Storybook initialization

- Added `--skip-install` flag to the `create-storybook` command in both `.circleci/config.yml` and `.circleci/src/jobs/test-init-features.yml` to prevent automatic package installation.
- Included an explicit `npm install` command to ensure dependencies are installed after Storybook setup.

valentinpalkovic committed:
Enhance error logging in BUNProxy and NPMProxy classes

- Updated the debug logging in both BUNProxy and NPMProxy to include the error message when an issue occurs while finding dependencies metadata using npm. This change improves the visibility of errors for better debugging.

valentinpalkovic committed:
Refactor spinner and task log functions to support non-TTY environments

- Added conditional handling in the `spinner` and `taskLog` functions to log messages using the `logger` when not in a TTY environment.
- Improved console.log patching for both spinner and task log methods to ensure consistent logging behavior across different environments.

valentinpalkovic committed:
Refactor terminal check for spinner and task log functions

- Replaced the `isTTY` constant with an `isInteractiveTerminal` function to improve the check for interactive terminal environments.
- Updated the `spinner` and `taskLog` functions to utilize the new terminal check, ensuring consistent behavior across different execution contexts.

valentinpalkovic committed:
Update CircleCI configuration to remove redundant npm install during Storybook initialization

- Removed the `npm install` command from both `.circleci/config.yml` and `.circleci/src/jobs/test-init-features.yml` as it is no longer necessary with the `--skip-install` flag in the `create-storybook` command.

valentinpalkovic committed:
Refactor Playwright installation logic in AddonVitestService

- Updated the `installPlaywright` method to use a `yes` option instead of `skipInstall` for better control over installation prompts.
- Enhanced user experience by providing a warning message when Playwright installation is skipped, along with instructions for manual installation.
- Modified the `AddonConfigurationCommand` to call `installPlaywright` with the new options, ensuring proper handling during addon configuration.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix type issues

valentinpalkovic committed:
Fix types

valentinpalkovic committed:
Fix error handling in executeTaskWithSpinner to stop task on SIGINT

valentinpalkovic committed:
Add nextjsToNextjsVite migration fix to allFixes export

valentinpalkovic committed:
Fix indentation handling in wrapTextForClackHint and update prompt-functions to pass default value

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Refactor UserPreferencesCommand to use CommandOptions for better configuration handling and remove installType from results

valentinpalkovic committed:
Delete integration test attempt

valentinpalkovic committed:
Enhance logging functions to conditionally log messages based on the 'info' level, improving performance and reducing unnecessary console output.

valentinpalkovic committed:
Refactor telemetry logging to remove unnecessary newline and ensure consistent message formatting; update telemetry call in project scaffolding to await for better async handling.

valentinpalkovic committed:
Various logging improvements

valentinpalkovic committed:
Refactor logging and task execution to improve message handling and reduce unnecessary output; introduce conditional logging for spinner and task messages.

valentinpalkovic committed:
Improve logging

valentinpalkovic committed:
Core: Move builder-specific mocking logic into builders

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Add caching for package.json files in JsPackageManager to optimize file system calls and improve performance; enhance logging for package.json operations.

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Minor logging fixes

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Update feature selection in GeneratorExecutionCommand to use selectedFeatures instead of options.features

valentinpalkovic committed:
Add support for Nuxt framework in get-storybook-info utility

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Refactor logging messages in AddonConfigurationCommand for clarity

valentinpalkovic committed:
Enhance logger function signatures for better type safety and clarity; refactor logging in AddonConfigurationCommand to log each addon result as a separate entry.

valentinpalkovic committed:
Add 'doctor' event type to telemetry and implement telemetry logging for the 'doctor' command in CLI. This enhances tracking and improves user feedback during the command execution.

valentinpalkovic committed:
Improve error handling and logging in dependency installation process; update messages for clarity and add manual instructions for addon configuration failures.

valentinpalkovic committed:
Refactor info logging in Vite logger to include spacing for improved readability; remove unused framework name import in Vite config.

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Improve error logging in extractMockCalls function to include error message details for better debugging.

valentinpalkovic committed:
Update framework type definitions to allow null values in various commands and interfaces; enhance dependency handling in React Native generator.

valentinpalkovic committed:
Refactor logging in FinalizationCommand to use warning level for debug log message; update intro logging to remove unnecessary newline for improved output formatting.

valentinpalkovic committed:
Several test and type fixes

valentinpalkovic committed:
Merge remote-tracking branch 'origin/valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Cleanup

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Remove obsolete outro's

valentinpalkovic committed:
Fix nextjs-vite automigration to update to the right version

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Refactor upgrade process to improve logging and error handling; streamline project selection and dependency updates.

valentinpalkovic committed:
Enhance upgrade logging by replacing log messages with step indicators for better clarity on upgrade status; improve user feedback on project upgrade results.

valentinpalkovic committed:
Refactor migration summary logging to improve readability by formatting URLs on separate lines; enhance clarity for users seeking migration guidance.

valentinpalkovic committed:
Add NX project detection error handling in ProjectDetectionCommand; throw NxProjectDetectedError for NX projects

valentinpalkovic committed:
Fix condition in ProjectDetectionCommand to correctly handle force option; ensure proper initialization flow for non-Angular projects.

valentinpalkovic committed:
Improve user prompt message formatting in ProjectDetectionCommand by using dedent for better readability; enhance clarity for users during initialization confirmation.

valentinpalkovic committed:
Cleanup logging new lines

valentinpalkovic committed:
Update '--write-logs' option description to specify log file name; add postAction hook for log writing in create-storybook command.

valentinpalkovic committed:
Enhance ProjectDetectionCommand to include an additional check for the 'yes' option, allowing users to bypass confirmation when initializing Storybook in existing projects; improve initialization flow for better user experience.

valentinpalkovic committed:
Merge remote-tracking branch 'origin/valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Refactor command execution across package managers to use a unified executeCommand utility; update related tests and improve command handling in various components for consistency and clarity.

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Refactor BUNProxy command execution to use 'npx' instead of 'bun run' for improved safety and reliability; added comments to clarify the change.

valentinpalkovic committed:
Fix: Update installDependencies method to use arrow function for improved context binding in JsPackageManager

valentinpalkovic committed:
Update CircleCI configuration to reduce parallelism for e2e-dev workflow from 29 to 28

valentinpalkovic committed:
Refactor package manager commands to replace runRemoteCommand with runPackageCommand for improved consistency across BUNProxy, NPMProxy, PNPMProxy, Yarn1Proxy, and Yarn2Proxy; update related tests accordingly.

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Remove abstract method

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Enhance project scaffolding by adding 'Other' option for unsupported frameworks; update warning message for clarity and exit process on selection.

valentinpalkovic committed:
Refactor command execution by removing 'shell: true' option in various spawn and execa calls for improved consistency and security across the codebase.

valentinpalkovic committed:
Update telemetry notification message to remove unnecessary formatting for improved clarity.

valentinpalkovic committed:
Remove unnecessary formatting from abort messages in task execution for improved clarity.

valentinpalkovic committed:
Refactor FinalizationCommand to remove selectedFeatures parameter, simplifying execution and logging process.

valentinpalkovic committed:
Remove unnecessary output from initiation completion logger for improved clarity.

valentinpalkovic committed:
Add silent flag for npm package manager in Storybook initiation to reduce output

valentinpalkovic committed:
Refactor logging output in various components for improved clarity and consistency; remove cli-table3 dependency.

valentinpalkovic committed:
Update Playwright installation prompt to indicate abort option; remove redundant abort message logging in task execution for clarity.

valentinpalkovic committed:
Update code/lib/cli-storybook/src/upgrade.ts

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

valentinpalkovic committed:
Refactor postinstall command execution to use spawnSync for synchronous execution; improve command argument handling.

valentinpalkovic committed:
Refactor sandbox function to remove unused borderColor variable and update border styling to use rounded property for improved clarity in output.

valentinpalkovic committed:
Fix command argument formatting in dispatcher.ts to remove unnecessary quotes around the index.js path for improved execution clarity.

valentinpalkovic committed:
Refactor postinstall command argument handling for improved clarity by removing unnecessary quotes and using array syntax.

valentinpalkovic committed:
Refactor command execution in link.ts to utilize executeCommand for improved clarity and consistency in handling git and yarn commands.

valentinpalkovic committed:
Remove 'skip-install' option from init function in sandbox-parts.ts for improved clarity in option handling.

ghengeveld commented on code/addons/vitest/src/postinstall.ts:
```diff
@@ -1,381 +1,200 @@
+    if (info.framework === SupportedFramework.NEXTJS) {
+      const allDeps = packageManager.getAllDependencies();
+      if (!allDeps['@storybook/nextjs-vite']) {
+        // TODO: Tell people to migrate first to nextjs-vite
```
Will you address this TODO?

ghengeveld commented on code/core/scripts/generate-source-files.ts:
```diff
@@ -95,16 +95,26 @@ async function generateFrameworksFile(prettierConfig: prettier.Options | null):
+  };
+
+  const coreFrameworks = readFrameworks.sort().map(formatFramework).join(',\n');
+  const communityFrameworks = thirdPartyFrameworks.map(formatFramework).join(',\n');
```
Just wondering, should this respect Windows linefeeds or is that handled elsewhere (maybe by Prettier)?

ghengeveld commented on code/core/src/bin/core.ts:
```diff
@@ -36,7 +36,10 @@ const command = (name: string) =>
-    .option('--write-logs', 'Write all debug logs to a file at the end of the run')
+    .option(
+      '--write-logs',
+      'Write all debug logs to the debug-storybook.log file at the end of the run'
```
Doesn't this flag accept a custom filename?

ghengeveld commented on code/core/src/core-server/mocking-utils/extract.ts:
```diff
@@ -186,7 +186,7 @@ export function extractMockCalls(
     return mocks;
   } catch (error) {
-    logger.debug('Error extracting mock calls', error);
+    logger.debug('Error extracting mock calls' + String(error));
```
```suggestion
    logger.debug('Error extracting mock calls: ' + String(error));
```

ghengeveld commented on code/core/src/core-server/build-static.ts:
```diff
@@ -70,7 +68,7 @@ export async function buildStaticStandalone(options: BuildStaticStandaloneOption
   );
 
-  logger.info('=> Loading presets');
+  logger.step('Loading presets');
```
How useful is this log really?

ghengeveld commented on code/core/src/core-server/withTelemetry.ts:
```diff
@@ -18,11 +25,12 @@ const promptCrashReports = async () => {
+  const enableCrashReports = await prompt.confirm({
+    message: dedent`
+      Send anonymous crash reports to help improve Storybook?
+      This helps us improve Storybook and fix bugs faster.
```
You're mentioning _help improve_ twice without any further details. Seems superfluous.

ghengeveld commented on code/core/src/node-logger/index.test.ts:
```diff
@@ -33,14 +36,13 @@ vi.mock('npmlog', () => ({
-  it('should have an info method', () => {
-    const message = 'information';
-    logger.info(message);
-    expect(npmlog.info).toHaveBeenCalledWith('', message);
```
You forgot to remove the import for `npmlog`.

ghengeveld commented on code/core/src/node-logger/tasks.ts:
```diff
@@ -1,60 +1,134 @@
     const errorMessage = err instanceof Error ? (err.stack ?? err.message) : String(err);
     logTracker.addLog('error', error, { error: errorMessage });
-    task.error(error);
+    log(CLI_COLORS.error(String((err as any).message ?? err)));
```
Why not use `errorMessage`? Unless you don't want the full stack here.

ghengeveld commented on code/core/src/telemetry/error-collector.ts:
```diff
@@ -0,0 +1,22 @@
+/** Service for collecting errors during Storybook initialization */
+export class ErrorCollector {
```
Could you add a docblock to explain why this class exists?

ghengeveld commented on code/lib/cli-storybook/src/automigrate/helpers/cleanLog.ts:
```diff
@@ -10,13 +8,3 @@ export const ansiRegex = ({ onlyFirst = false } = {}) => {
 
   return new RegExp(pattern, onlyFirst ? undefined : 'g');
 };
```
I think we don't use `ansiRegex` anywhere else, so you can probably remove the entire file.

ghengeveld commented on code/lib/cli-storybook/src/automigrate/index.ts:
```diff
@@ -164,7 +172,7 @@ export const automigrate = async ({
   }
 
-  logger.log('🔎 checking possible migrations..');
+  logger.step('Checking possible migrations..');
```
Maybe remove the emojis in the other logs as well.

valentinpalkovic committed:
Add debug logging for version handling in postinstall and package manager classes for improved traceability

valentinpalkovic committed:
Enhance version handling in JsPackageManager by incorporating version coercion for improved accuracy in installed version reporting.

valentinpalkovic committed:
Refactor JsPackageManager tests to use spyOn for mocking latestVersion method, enhancing test accuracy and clarity.

ghengeveld commented on code/lib/cli-storybook/src/sandbox.ts:
```diff
@@ -197,13 +197,9 @@ export const sandbox = async ({
-        force: true,
-        dir: templateDestination,
-      });
+      spawnSync('npx', ['gitpick', gitPath, templateDestination, '-o']);
```
I would feel safer is this was a pinned-down version of gitpick. Supply chain attacks are a thing...
Also, this is a hidden dependency, not made explicit though package.json or node_modules. I don't love that.

ghengeveld replied:
I see we had `unjs/giget` before (you removed it). Why not use that instead of gitpick? They seem to do the same thing.

ghengeveld commented on code/lib/cli-storybook/src/sandbox.ts:
```diff
@@ -197,13 +197,9 @@ export const sandbox = async ({
+      spawnSync('npx', ['gitpick', gitPath, templateDestination, '-o']);
       // throw an error if templateDestination is an empty directory
       if ((await readdir(templateDestination)).length === 0) {
         const selected = picocolors.yellow(templateId);
```
This is probably legit.

ghengeveld commented on code/lib/create-storybook/src/bin/run.ts:
```diff
@@ -47,7 +48,32 @@ const createStorybookProgram = program
-  );
+  )
+  .option(
+    '--write-logs',
```
Maybe this should be `--logfile` and accept a filepath (besides empty string for the default filename) so that it can be written elsewhere?

ghengeveld commented on code/lib/create-storybook/src/generators/REACT_NATIVE/index.ts:
```diff
@@ -1,71 +1,104 @@
+  },
+  postConfigure: ({ packageManager }) => {
+    logger.log(dedent`
+      ${CLI_COLORS.warning('React Native (RN) Storybook installation is not 100% automated.')}
```
I think we call it "Storybook for React Native" usually. Also I would just spell it out in full each time rather than use an acronym.

ghengeveld commented on code/lib/create-storybook/src/generators/REACT_SCRIPTS/index.ts:
```diff
@@ -2,69 +2,66 @@ import { existsSync } from 'node:fs';
-  const extraAddons = [`@storybook/preset-create-react-app`];
+    // TODO: Evaluate if adding prop-types is correct after removing pnp compatibility code in SB11
+    // Miscellaneous dependency to add to be sure Storybook + CRA is working fine with Yarn PnP mode
+    const extraPackages = ['webpack', 'prop-types'];
```
Does anyone still use prop-types for new projects? I guess maybe we're just keeping this around for sandboxes?

ghengeveld commented on code/lib/create-storybook/src/dependency-collector.ts:
```diff
@@ -0,0 +1,187 @@
+import type { JsPackageManager } from 'storybook/internal/common';
```
Unused type import.

ghengeveld commented on code/lib/create-storybook/src/initiate.ts:
````diff
@@ -1,812 +1,158 @@
- * ```
+ * This is a clean, command-based orchestration that replaces the monolithic 986-line implementation
+ * with a modular, testable approach.
  */
````
👏

ghengeveld commented on code/lib/create-storybook/src/scaffold-new-project.ts:
```diff
@@ -2,15 +2,12 @@ import { readdirSync } from 'node:fs';
 
 import type { PackageManagerName } from 'storybook/internal/common';
-import { logger } from 'storybook/internal/node-logger';
+import { logger, prompt } from 'storybook/internal/node-logger';
```
Unused import.

ghengeveld commented on code/lib/create-storybook/src/scaffold-new-project.ts:
```diff
@@ -118,59 +115,35 @@ export const scaffoldNewProject = async (
+    projectStrategy = await prompt.select({
+      message: dedent`
+        Empty directory detected:
+        Would you like to generate a new project from the following list?
```
```suggestion
        Empty directory detected. Would you like to generate a new project from the following list?
```

valentinpalkovic commented on code/core/src/core-server/build-dev.ts:
```diff
@@ -68,11 +67,12 @@ export async function buildDevStandalone(
+    const shouldChangePort = await prompt.confirm({
+      message: dedent`
+        Port ${options.port} is not available. 
+        Would you like to run Storybook on port ${port} instead?
```
A feature, which makes sense, but not as part of this PR. It is already quite big.

valentinpalkovic committed:
Update CircleCI configuration to remove loglevel flag from create-storybook command for cleaner output

valentinpalkovic committed:
Update CircleCI configuration to change executor class from medium+ to small and remove 'dev' feature from create-storybook command for streamlined setup

valentinpalkovic committed:
Enhance logging in logger.ts to counteract default behavior of clack prompt library and improve clarity in postinstall.ts error messages regarding package incompatibilities.

valentinpalkovic committed:
Add warning suppression to customViteLogger in logger.ts to prevent duplicate warnings

valentinpalkovic committed:
Small cleanup

valentinpalkovic committed:
Add test

valentinpalkovic committed:
Refactor package manager classes to remove synchronous command execution methods from BUNProxy, NPMProxy, PNPMProxy, Yarn1Proxy, and Yarn2Proxy for improved consistency and simplicity.

valentinpalkovic committed:
Update framework-to-renderer and framework-to-builder mappings to correctly associate NUXT, QWIK, and SOLID with VUE3 and VITE for improved framework compatibility.

valentinpalkovic committed:
Refactor getStorybookInfo function to ensure correct configuration directory usage and include version parameter for improved Storybook info retrieval.

valentinpalkovic committed:
Cleanup and small improvements

valentinpalkovic committed:
Refactor logger functions to use parameterized types for improved type safety and consistency in logging behavior.

valentinpalkovic committed:
Refactor runCompodoc to use asynchronous package command execution for improved performance and error handling.

valentinpalkovic committed:
Add comments

valentinpalkovic committed:
Update gitpick version in sandbox.ts to ensure compatibility and prevent errors on Windows

valentinpalkovic committed:
Refactor logger color constants to improve clarity and consistency in logging output.

valentinpalkovic committed:
Remove unused file

valentinpalkovic committed:
Implement dry run option for package.json updates in nextjs-to-nextjs-vite migration script to enhance user experience during migration.

valentinpalkovic committed:
Small cleanup

valentinpalkovic committed:
Refactor logging options in CLI commands to use --logfile for specifying log file paths, enhancing flexibility and consistency in log management.

valentinpalkovic committed:
Enhance logging functionality in Storybook initialization by passing logfile option to finalization steps, improving error tracking and user feedback.

valentinpalkovic committed:
Improve error handling in AddonConfigurationCommand by enhancing logging for unexpected errors during addon configuration.

valentinpalkovic committed:
Remove unused addon dependency functions for a11y and vitest, streamlining the DependencyInstallationCommand by directly utilizing AddonVitestService for dependency collection.

valentinpalkovic committed:
Refactor PreflightCheckCommand to use options directly for package manager and empty directory checks, improving code clarity and consistency.

valentinpalkovic committed:
Improve error handling in ProjectDetectionCommand by allowing specific errors to be rethrown, enhancing error management and logging clarity.

valentinpalkovic committed:
Small refactorings

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Remove --logfile option from CLI dev command, simplifying command options and improving clarity.

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Fix types

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Merge branch 'next' into valentin/cli-init-rework

valentinpalkovic committed:
Refactor JsPackageManager to normalize paths using resolve() for consistent cache keys across platforms

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Update Yarn1Proxy to correctly pass command arguments with '--' separator in executeCommand

valentinpalkovic committed:
Change default return value in getErrorLevel function from 'full' to 'error' when presetOptions is not provided

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-rework-design-adjustments

valentinpalkovic committed:
Refactor command execution in dispatcher and add synchronous command execution utility

valentinpalkovic committed:
Refactor taskLog implementation to simplify logging conditions and remove redundant checks for log levels

valentinpalkovic committed:
Refactor command spawning in dispatcher to separate command and arguments for improved clarity

valentinpalkovic committed:
Refactor telemetry error handling to conditionally send errors only when telemetry is enabled

valentinpalkovic committed:
Update stdio configuration in AddonVitestService to inherit and pipe for improved command output handling

valentinpalkovic committed:
Update Playwright installation command in AddonVitestService to use 'npx' for improved clarity

valentinpalkovic committed:
Update automigration logging to handle optional fixId for improved clarity

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Update yarn.lock file

valentinpalkovic committed:
Add RSBuild support

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-docs-performance

valentinpalkovic committed:
Merge pull request #32921 from storybookjs/valentin/cli-init-docs-performance

Core: Decouple Mocking Logic from Core into Builders

valentinpalkovic committed:
Merge branch 'valentin/cli-init-rework' into valentin/cli-init-rework-design-adjustments

valentinpalkovic committed:
Merge pull request #32984 from storybookjs/valentin/cli-init-rework-design-adjustments

CLI: Implement design feedback

valentinpalkovic committed:
Refactor package manager handling to use enums for type safety

- Updated `JsPackageManager` and its proxies to utilize `PackageManagerName` enum instead of string literals for package manager types.
- Adjusted related tests and implementations to reflect these changes, ensuring consistent usage across the codebase.
- Improved clarity and maintainability of package manager type definitions.

valentinpalkovic committed:
Refactor getApplicationFilesCountUncached to simplify command construction

- Removed unnecessary quotes around glob patterns in the `getApplicationFilesCountUncached` function.
- Changed the return statement to directly call `execCommandCountLines` without awaiting, as it is not necessary in this context.
- Improved code readability and performance by streamlining the command execution process.

valentinpalkovic committed:
Refactor JsPackageManagerFactory tests to use executeCommandSync

- Replaced `spawnSync` with `executeCommandSync` for command execution in `JsPackageManagerFactory.test.ts`.
- Updated mock implementations to align with the new command execution method, improving error handling and test reliability.
- Enhanced clarity in test cases by using structured options for command execution.

valentinpalkovic committed:
Refactor project type handling for improved consistency and type safety

valentinpalkovic committed:
Reorganize logic for better structure and readability

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Update configuration call in React Native generator to include context parameter for improved functionality

valentinpalkovic committed:
Import ProjectType from core for improved type handling in sandbox templates

valentinpalkovic committed:
Refactor import statements in sandbox templates for improved clarity and type handling

valentinpalkovic committed:
Update CircleCI configuration to remove 'dev' feature from Storybook initialization for streamlined setup

valentinpalkovic committed:
Update ProjectTypeService instantiation to handle package manager types more flexibly

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix preset loading in yarn pnp mode

valentinpalkovic committed:
Update taskLog function to require both interactive terminal and info logging for improved logging behavior

valentinpalkovic committed:
Refactor execCommandCountLines to accept command arguments and update related calls for improved flexibility

valentinpalkovic committed:
Enhance command execution by adding cross-platform support for Node-based CLI commands through a new resolveCommand function

valentinpalkovic committed:
Removed `execa` from `create-storybook`

valentinpalkovic committed:
Remove `execa` across storybook packages

valentinpalkovic committed:
Add 'ng' to command resolution and refactor command execution in Storybook

valentinpalkovic committed:
Cleanup

valentinpalkovic committed:
Refactor logging in spinner functions to use conditional logging based on log level

valentinpalkovic committed:
Cleanup

valentinpalkovic committed:
Update getApplicationFilesCountUncached to await execCommandCountLines for proper asynchronous handling

valentinpalkovic committed:
Update yarn.lock

valentinpalkovic committed:
Add shell option to command execution in scaffoldNewProject

valentinpalkovic committed:
Refactor command execution in dispatcher to use internal executeCommand and executeNodeCommand functions

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Fix tests

valentinpalkovic committed:
Enhance command execution in dispatcher by adding stdio inheritance for child processes

valentinpalkovic committed:
Refactor PackageManagerName enum to use 'yarn' instead of 'yarn1' and update PreflightCheckCommand to reflect this change. Add tests for FrameworkDetectionService to ensure proper framework and builder detection.

valentinpalkovic committed:
Shorten prompt

valentinpalkovic committed:
Addon-A11y: Fix a11y postinstall

valentinpalkovic committed:
Refactor a11y postinstall to pass args as an object in runPackageCommand

valentinpalkovic committed:
Refactor upgrade function to use PackageManagerName enum for better clarity and maintainability

valentinpalkovic committed:
Refactor UserPreferencesCommand and ProjectDetectionCommand to use Array<Feature> for features and update CommandOptions interface accordingly

valentinpalkovic committed:
Fix tests and linting

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Linting

valentinpalkovic committed:
Update CLI color configuration to use cyan for info messages on Windows platform

valentinpalkovic committed:
Fix exit code handling in dispatcher to default to 1 if undefined

valentinpalkovic committed:
Refactor GeneratorExecutionCommand to use type assertion with 'satisfies' for GeneratorOptions and ensure boolean values for options

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Add error names to AddonVitestPostinstall and AutomigrateError classes for better error identification

valentinpalkovic committed:
Remove comment

valentinpalkovic committed:
Refactor GeneratorExecutionCommand to use type assertion with 'as' for GeneratorOptions

valentinpalkovic committed:
Introduce --no-features flag for init command

valentinpalkovic committed:
Remove eslint-disable-next-line comments

valentinpalkovic committed:
Remove comments

valentinpalkovic committed:
Adjust copy

valentinpalkovic committed:
Sort supported frameworks

valentinpalkovic committed:
Sort project type

valentinpalkovic committed:
Sort third party frameworks

valentinpalkovic committed:
Return error object key instead of an errors array for installPlaywright method

valentinpalkovic committed:
Change executor class from medium+ to small for test-init-features job

valentinpalkovic committed:
Fix version of @clack/prompts

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Sort community frameworks

valentinpalkovic committed:
FIx bug related to --skip-install being ignored in AddonConfigurationCommand

valentinpalkovic committed:
Merge branch 'next' into valentin/cli-init-rework

valentinpalkovic committed:
Cleanup

valentinpalkovic committed:
Add tests for project type service and adjust error messaging

valentinpalkovic committed:
Remove debug log for features in run.ts

valentinpalkovic committed:
Refactor telemetry notification logging to use info level and remove unused CLI_COLORS import

valentinpalkovic committed:
Update configuration messages to reflect file extension based on language and improve version satisfaction checks in ProjectTypeService

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/cli-init-rework

valentinpalkovic committed:
Update yarn lock

valentinpalkovic committed:
Revert "Remove eslint-disable-next-line comments"

This reverts commit 0abdc8c2b2b6e372b1c01dbfef9ce7ee28936160.
